linera-core/src/join_set_ext.rs

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) Zefchain Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//! An extension trait to allow determining at compile time how tasks are spawned on the Tokio
+//! runtime.
+//!
+//! In most cases the [`Future`] task to be spawned should implement [`Send`], but that's
+//! not possible when compiling for `wasm32-unknown-unknown`. In that case, the task is
+//! spawned in a [`LocalSet`][`tokio::tast::LocalSet`].
+
+use std::future::Future;
+
+use tokio::{
+    sync::oneshot,
+    task::{AbortHandle, JoinSet},
+};
+
+/// An extension trait for the [`JoinSet`] type.
+#[cfg(not(target_arch = "wasm32"))]
+pub trait JoinSetExt: Sized {
+    /// Spawns a `future` task on this [`JoinSet`] using [`JoinSet::spawn`].
+    ///
+    /// Returns a [`oneshot::Receiver`] to receive the `future`'s output, and an
+    /// [`AbortHandle`] to cancel execution of the task.
+    fn spawn_task<F>(&mut self, future: F) -> (oneshot::Receiver<F::Output>, AbortHandle)
+    where
+        F: Future + Send + 'static,
+        F::Output: Send;
+
+    /// Awaits all tasks spawned in this [`JoinSet`].
+    fn await_all_tasks(&mut self) -> impl Future<Output = ()> + Send;
+
+    /// Reaps tasks that have finished.
+    fn reap_finished_tasks(&mut self);
+}
+
+/// An extension trait for the [`JoinSet`] type.
+#[cfg(target_arch = "wasm32")]
+pub trait JoinSetExt: Sized {
+    /// Spawns a `future` task on this [`JoinSet`] using [`JoinSet::spawn_local`].
+    ///
+    /// Returns a [`oneshot::Receiver`] to receive the `future`'s output, and an
+    /// [`AbortHandle`] to cancel execution of the task.
+    fn spawn_task<F>(&mut self, future: F) -> (oneshot::Receiver<F::Output>, AbortHandle)
```

Comment:
I think that `(oneshot::Receiver<F::Output>, AbortHandle)` is more clearly written as `AbortOnDrop<impl Future<Output = F::Output>>`.

Reply:
I like the idea, but I think it changes the semantics a bit. The problem is that it becomes a bit redundant. The `JoinSet` is already an "abort all tasks on drop", and the `AbortHandle` is more a preemptive measure. I think we can do this when revisiting the entire `JoinSetExt` mechanism.

In the meantime, I refactored to create a `TaskHandle` type instead, which we can implement `Drop` later on if needed :+1:

Reply:
I think merely returning `impl Future<Output = F::Output>` is already sufficient to have a ‘task handle’ that can be dropped — every type may have a `Drop` implementation, and a `Future` in Rust is precisely a ‘task handle’ that is referring to the (future) output of a task.  I left the `AbortOnDrop` in my suggestion explicitly (even though it's not entirely needed) because it indicates clearly to the consumer that dropping causes the task to be aborted, rather than joined or detached.

But I'm not going to block on this, we can discuss it later :)